src/components/TodoItems.tsx

```diff
@@ -0,0 +1,198 @@
+import classNames from 'classnames';
+import React, { useEffect, useRef, useState } from 'react';
+import { deleteTodo, updateTodo } from '../api/todos';
+import { Todo } from '../types/Todo';
+import { ErrorList } from '../types/ErrorList';
+import { useTodosContext } from '../context/TodoContext';
+
+interface Props {
+  todo: Todo;
+}
+
+export const TodoItem: React.FC<Props> = ({ todo }) => {
+  const {
+    processingIds,
+    focusInput,
+    todos,
+    setTodos,
+    handleError,
+    handleDeleteTodo,
+  } = useTodosContext();
+
+  const [queryEditing, setQueryEditing] = useState(todo.title);
+  const [isLoading, setIsLoading] = useState(false);
+  const [isEditing, setIsEditing] = useState(false);
+
+  const inputTodoRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    focusInput();
+    if (isEditing && inputTodoRef.current) {
+      inputTodoRef.current?.focus();
+    }
+  }, [isEditing]);
+
+  const handleChecker = async (ItemID: number) => {
+    const todosChecker = todos.map(prevItem => {
+      if (prevItem.id === ItemID) {
+        return {
+          ...prevItem,
+          completed: !prevItem.completed,
+        };
+      }
+
+      return prevItem;
+    });
+
+    setIsLoading(true);
+
+    const newTodo = {
+      ...todo,
+      completed: !todo.completed,
+    };
+
+    await updateTodo(ItemID, newTodo)
+      .then(() => {
+        setTodos(todosChecker);
+      })
+      .catch(() => {
+        handleError(ErrorList.UpdateTodo);
+      })
+      .finally(() => {
+        focusInput();
+        setIsLoading(false);
+      });
+  };
+
+  const saveInputValue = () => {
+    if (queryEditing.trim()) {
+      setIsLoading(true);
+      if (queryEditing.trim() === todo.title) {
+        setIsEditing(false);
+        focusInput();
+        setIsLoading(false);
+      } else {
+        const newTodo = {
+          ...todo,
+          title: queryEditing.trim(),
+        };
+
+        updateTodo(todo.id, newTodo)
+          .then(todoValue => {
+            setTodos(prevTodos =>
+              prevTodos.map(currentTodo =>
+                currentTodo.id === todo.id ? todoValue : currentTodo,
+              ),
+            );
+            setIsEditing(false);
+            focusInput();
+            setQueryEditing(queryEditing.trim());
+          })
+          .catch(() => {
+            handleError(ErrorList.UpdateTodo);
+          })
+          .finally(() => {
+            setIsLoading(false);
+          });
+      }
+    } else {
+      setIsLoading(true);
+      deleteTodo(todo.id)
+        .then(() => {
+          setTodos(todos.filter(item => item.id !== todo.id));
+          setIsEditing(false);
+          focusInput();
+        })
+        .catch(() => {
+          handleError(ErrorList.DeleteTodo);
+        })
+        .finally(() => {
+          setIsLoading(false);
+        });
+    }
+  };
```

Comment:
This logic looks a little bit messy, try to think how we can reduce nesting here